execution_engine/src/runtime/externals.rs

```diff
@@ -1106,6 +1107,85 @@ where
+                    &host_function_costs.manage_message_topic,
+                    [
+                        topic_name_ptr,
+                        topic_name_size,
+                        operation_ptr,
+                        operation_size,
+                    ],
+                )?;
+
+                let limits = self.context.engine_config().wasm_config().messages_limits();
+
+                if topic_name_size > limits.max_topic_name_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MaxTopicNameSizeExceeded,
+                    )))));
+                }
+
+                let topic_name = self.string_from_mem(topic_name_ptr, topic_name_size)?;
+
+                if operation_size as usize > MessageTopicOperation::max_serialized_len() {
+                    return Err(Trap::from(Error::InvalidMessageTopicOperation));
+                }
+                let topic_operation = self
+                    .t_from_mem(operation_ptr, operation_size)
+                    .map_err(|_e| Trap::from(Error::InvalidMessageTopicOperation))?;
+
+                // don't allow managing topics for accounts.
+                if let Key::Account(_) = self.context.get_entity_address() {
+                    return Err(Trap::from(Error::InvalidContext));
+                }
+
+                let result = match topic_operation {
+                    MessageTopicOperation::Add => {
+                        self.add_message_topic(topic_name).map_err(Trap::from)?
+                    }
+                };
+
+                Ok(Some(RuntimeValue::I32(api_error::i32_from(result))))
+            }
+            FunctionIndex::EmitMessage => {
+                // args(0) = pointer to the serialized topic name string in wasm memory
+                // args(1) = size of the serialized name string in wasm memory
+                // args(2) = pointer to the serialized message payload in wasm memory
+                // args(3) = size of the serialized message payload in wasm memory
+                let (topic_name_ptr, topic_name_size, message_ptr, message_size) =
+                    Args::parse(args)?;
+                self.charge_host_function_call(
+                    &host_function_costs.emit_message,
+                    [topic_name_ptr, topic_name_size, message_ptr, message_size],
+                )?;
+
+                let limits = self.context.engine_config().wasm_config().messages_limits();
+
+                if topic_name_size > limits.max_topic_name_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MaxTopicNameSizeExceeded,
+                    )))));
+                }
+
+                if message_size > limits.max_message_size() {
+                    return Ok(Some(RuntimeValue::I32(api_error::i32_from(Err(
+                        ApiError::MessageTooLarge,
+                    )))));
+                }
```

Comment:
Please consider extracting the validation part to a function and cover it with UT.